core/src/apps/cardano/address.py

```diff
@@ -1,145 +1,350 @@
+        return bech32.encode(hrp, address)
+    else:
+        raise ValueError
+
+
+def derive_address_bytes(
+    keychain: seed.Keychain,
+    parameters: CardanoAddressParametersType,
+    protocol_magic: int,
+    network_id: int,
+) -> bytes:
+    is_byron_address = parameters.address_type == CardanoAddressType.BYRON
+
+    if is_byron_address:
+        address = _derive_byron_address(keychain, parameters.address_n, protocol_magic)
+    else:
+        address = _derive_shelley_address(keychain, parameters, network_id)
 
+    return address
 
-def validate_full_path(path: list) -> bool:
+
+def _derive_byron_address(
+    keychain: seed.Keychain, path: List[int], protocol_magic: int
+) -> bytes:
+    if not is_byron_path(path):
+        raise wire.DataError("Invalid path for byron address!")
+
+    address = derive_byron_address(keychain, path, protocol_magic)
+    return address
+
+
+def _derive_shelley_address(
+    keychain: seed.Keychain, parameters: CardanoAddressParametersType, network_id: int,
+) -> bytes:
+    if not is_shelley_path(parameters.address_n):
+        raise wire.DataError("Invalid path for shelley address!")
+
+    if parameters.address_type == CardanoAddressType.BASE:
+        address = _derive_base_address(
+            keychain,
+            parameters.address_n,
+            parameters.address_n_staking,
+            parameters.staking_key_hash,
+            network_id,
+        )
+    elif parameters.address_type == CardanoAddressType.ENTERPRISE:
+        address = _derive_enterprise_address(keychain, parameters.address_n, network_id)
+    elif parameters.address_type == CardanoAddressType.POINTER:
+        address = _derive_pointer_address(
+            keychain, parameters.address_n, parameters.certificate_pointer, network_id,
+        )
+    elif parameters.address_type == CardanoAddressType.REWARD:
+        address = _derive_reward_address(keychain, parameters.address_n, network_id)
+    else:
+        raise ValueError
+
+    return address
+
+
+def _create_address_header(
+    address_type: EnumTypeCardanoAddressType, network_id: int
+) -> bytes:
+    header = address_type << 4 | network_id
+    return header.to_bytes(1, "little")
+
+
+def _derive_base_address(
+    keychain: seed.Keychain,
+    path: List[int],
+    staking_path: List[int],
+    staking_key_hash: bytes,
+    network_id: int,
+) -> bytes:
+    header = _create_address_header(CardanoAddressType.BASE, network_id)
+    spending_key_hash = get_public_key_hash(keychain, path)
+
+    _validate_base_address_staking_info(staking_path, staking_key_hash)
+
+    if staking_key_hash is None:
+        staking_key_hash = get_public_key_hash(keychain, staking_path)
+
+    return header + spending_key_hash + staking_key_hash
+
+
+def _validate_base_address_staking_info(
+    staking_path: List[int], staking_key_hash: bytes,
+) -> None:
+    if staking_key_hash is None and staking_path is None:
```

Comment:
istm we can accomplish what we want by this:
```
if (staking_key_hash is None) == (staking_path is None):
    raise wire.DataError("Base address needs either a staking path or a staking key hash!")
```

Reply:
Hm, that doesn't seem to work. It turns out the original condition was also wrong, since `staking_path` can never be none, only empty. So what do you think of this:
```
if (staking_key_hash is None) == (len(staking_path) == 0):
    raise wire.DataError("Base address needs either a staking path or a staking key hash!")
```
or the `len(staking_path) == 0` can be replaced by `not bool(staking_path)`. Or I can simply make a second check.

Reply:
you can use `not staking_path` without the `bool()`

but now i'm wondering if the other checks also check for `None`?

Reply:
...and the other checks look correct

Reply:
What other checks do you mean? ... I checked and there seems to be no more path checking against None.

Reply:
i meant the subsequent checks on the staking path (in layout, sign_tx). but as you say, these seem OK

Reply:
Updated [here](https://github.com/trezor/trezor-firmware/pull/1112/commits/4a6b891da4374598daefa51fee99f30cfdae6852).